File: SOLID/code-examples/chapter-08.cpp
// FILE: chapter-08.cpp
// Chapter 8 — Liskov Substitution Principle — C++

#include <iostream>
#include <vector>
#include <memory>
#include <stdexcept>
using namespace std;

// ❌ BAD: Violates LSP - Square changes Rectangle behavior
class BadRectangle {
protected:
    double width, height;
public:
    BadRectangle(double w, double h) : width(w), height(h) {}
    
    virtual void setWidth(double w) { width = w; }
    virtual void setHeight(double h) { height = h; }
    
    double getWidth() const { return width; }
    double getHeight() const { return height; }
    double getArea() const { return width * height; }
};

class BadSquare : public BadRectangle {
public:
    BadSquare(double side) : BadRectangle(side, side) {}
    
    void setWidth(double w) override {
        width = height = w;  // Changes both dimensions!
    }
    
    void setHeight(double h) override {
        width = height = h;  // Changes both dimensions!
    }
};

// ✅ GOOD: Follows LSP - proper abstraction
class Shape {
public:
    virtual double getArea() const = 0;
    virtual void scale(double factor) = 0;
    virtual ~Shape() = default;
};

class Rectangle : public Shape {
protected:
    double width, height;
public:
    Rectangle(double w, double h) : width(w), height(h) {}
    
    double getArea() const override {
        return width * height;
    }
    
    void scale(double factor) override {
        width *= factor;
        height *= factor;
    }
    
    double getWidth() const { return width; }
    double getHeight() const { return height; }
};

class Square : public Shape {
private:
    double side;
public:
    Square(double s) : side(s) {}
    
    double getArea() const override {
        return side * side;
    }
    
    void scale(double factor) override {
        side *= factor;
    }
    
    double getSide() const { return side; }
};

// Function that works with any Shape (LSP compliant)
void processShapes(const vector<unique_ptr<Shape>>& shapes) {
    for (const auto& shape : shapes) {
        cout << "Shape area: " << shape->getArea() << endl;
        // Can safely call any Shape method on any derived class
    }
}

int main() {
    cout << "🔄 Liskov Substitution Principle (C++)" << endl;
    cout << "=======================================" << endl << endl;
    
    cout << "❌ Bad example (violates LSP):" << endl;
    BadRectangle* rect = new BadSquare(5);
    rect->setWidth(10);
    rect->setHeight(8);
    cout << "Expected area: 80, Actual area: " << rect->getArea() << endl;
    delete rect;
    
    cout << endl << "✅ Good example (follows LSP):" << endl;
    vector<unique_ptr<Shape>> shapes;
    shapes.push_back(make_unique<Rectangle>(4, 6));
    shapes.push_back(make_unique<Square>(5));
    
    processShapes(shapes);
    
    cout << endl << "💡 LSP Benefits:" << endl;
    cout << "   ✓ Derived classes are truly substitutable" << endl;
    cout << "   ✓ No unexpected behavior changes" << endl;
    cout << "   ✓ Polymorphism works correctly" << endl;
    
    return 0;
}

File: SOLID/code-examples/chapter-08.py
# FILE: chapter-08.py
# Chapter 8 — Liskov Substitution Principle — Python

from abc import ABC, abstractmethod

# ❌ BAD: Violates LSP - Square changes Rectangle behavior
class BadRectangle:
    def __init__(self, width, height):
        self._width = width
        self._height = height
    
    def set_width(self, width):
        self._width = width
    
    def set_height(self, height):
        self._height = height
    
    def get_width(self):
        return self._width
    
    def get_height(self):
        return self._height
    
    def get_area(self):
        return self._width * self._height

class BadSquare(BadRectangle):
    def __init__(self, side):
        super().__init__(side, side)
    
    def set_width(self, width):
        self._width = self._height = width  # Changes both dimensions!
    
    def set_height(self, height):
        self._width = self._height = height  # Changes both dimensions!

# ✅ GOOD: Follows LSP - proper abstraction
class Shape(ABC):
    @abstractmethod
    def get_area(self):
        pass
    
    @abstractmethod
    def scale(self, factor):
        pass

class Rectangle(Shape):
    def __init__(self, width, height):
        self._width = width
        self._height = height
    
    def get_area(self):
        return self._width * self._height
    
    def scale(self, factor):
        self._width *= factor
        self._height *= factor
    
    def get_width(self):
        return self._width
    
    def get_height(self):
        return self._height

class Square(Shape):
    def __init__(self, side):
        self._side = side
    
    def get_area(self):
        return self._side ** 2
    
    def scale(self, factor):
        self._side *= factor
    
    def get_side(self):
        return self._side

# Function that works with any Shape (LSP compliant)
def process_shapes(shapes):
    for shape in shapes:
        print(f"Shape area: {shape.get_area()}")
        # Can safely call any Shape method on any derived class

def main():
    print("🔄 Liskov Substitution Principle (Python)")
    print("==========================================\n")
    
    print("❌ Bad example (violates LSP):")
    rect = BadSquare(5)  # Treated as BadRectangle
    rect.set_width(10)
    rect.set_height(8)
    print(f"Expected area: 80, Actual area: {rect.get_area()}")
    
    print("\n✅ Good example (follows LSP):")
    shapes = [
        Rectangle(4, 6),
        Square(5)
    ]
    
    process_shapes(shapes)
    
    print("\n💡 LSP Benefits:")
    print("   ✓ Derived classes are truly substitutable")
    print("   ✓ No unexpected behavior changes")
    print("   ✓ Polymorphism works correctly")

if __name__ == "__main__":
    main()